pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -201,66 +251,119 @@ func (cfg Config) toAuthOptions() gophercloud.AuthOptions {
 
 func (cfg Config) toAuth3Options() tokens3.AuthOptions {
 	return tokens3.AuthOptions{
-		IdentityEndpoint: cfg.Global.AuthURL,
-		Username:         cfg.Global.Username,
-		UserID:           cfg.Global.UserID,
-		Password:         cfg.Global.Password,
-		DomainID:         cfg.Global.DomainID,
-		DomainName:       cfg.Global.DomainName,
+		IdentityEndpoint: cfg.Global.AuthInfo.AuthURL,
+		Username:         cfg.Global.AuthInfo.Username,
+		UserID:           cfg.Global.AuthInfo.UserID,
+		Password:         cfg.Global.AuthInfo.Password,
+		DomainID:         cfg.Global.AuthInfo.UserDomainID,
+		DomainName:       cfg.Global.AuthInfo.UserDomainName,
 		AllowReauth:      true,
 	}
 }
 
 // configFromEnv allows setting up credentials etc using the
 // standard OS_* OpenStack client environment variables.
 func configFromEnv() (cfg Config, ok bool) {
```

Comment:
We may be able to simplify this function further by relying completely on clientconfig to read most of the env variables.